Add prompts section to schematic guide (#32728)
angular/angular: aio/content/guide/schematics-authoring.md

| :----------- | :-------------------|
| confirmation | A yes or no question; ideal for Boolean options. |
| input | Textual input; ideal for string or number options. |
| list | A predefined set of allowed values. |

In the short form, the type is inferred from the property's type and constraints.

| Property Schema |	Prompt Type |
| :--------------- | :------------- |
| "type": "boolean" |	confirmation ("yes"=`true`, "no"=`false`) |
| "type": "string"  |	input |
| "type": "number"  |	input (only valid numbers accepted) |
| "type": "integer" |	input (only valid numbers accepted) |
| "enum": [...]   	| list 	(enum members become list selections) |

The prompt runtime automatically validates the provided response against the constraints provided in the JSON schema.
If the value is not acceptable, the user is prompted for a new value.
This ensures that any values passed to the schematic meet the expectations of the schematic's implementation, so that you do not need to add additional checks within the schematic's code.

#### Prompt long-form syntax

The `x-prompt` field syntax supports a long form for cases where you require additional customization and control over the prompt.

> In the this form, the `x-prompt` field value is a JSON object with subfields that customize the behavior of the prompt.

Reviewer: "In ~~the~~ this form,"

In most cases, this is all that is required.
Notice however, that the two prompts expect different types of input.
When using the shorthand form, the most appropriate type is automatically selected based on the property's schema.
In the example, the `name` prompt uses the `input` type because it it is a string property.
The `useColor` prompt use a `confirmation` type because it is a Boolean property.
In this case, "yes" corresponds to `true` and "no" corresponds to `false`.

There are three supported input types.

| Input type | Description |
| :----------- | :-------------------|
| confirmation | A yes or no question; ideal for Boolean options. |
| input | Textual input; ideal for string or number options. |
| list | A predefined set of allowed values. |

In the short form, the type is inferred from the property's type and constraints.

| Property Schema |	Prompt Type |
| :--------------- | :------------- |
| "type": "boolean" |	confirmation ("yes"=`true`, "no"=`false`) |
| "type": "string"  |	input |
| "type": "number"  |	input (only valid numbers accepted) |
| "type": "integer" |	input (only valid numbers accepted) |

> | "enum": [...]   	| list 	(enum members become list selections) |

Reviewer 2: I think that there could be some value in showing another example that uses the short form with `enum`. When I read through this, that feels like the biggest gap where I have questions in my head about how it should/might work.

Reviewer: You could use the existing list example but just use `"x-prompt": "Which stylesheet format would you like to use?"` instead.  The enum values become the displayed values. There isn't any usage in the Angular schematics.  There is a multi-select list ([link](https://github.com/angular/angular-cli/blob/8ea892cb7ea124127fa3867a8c3edd44e855106f/packages/schematics/angular/guard/schema.json#L53)) but that's newer and not at all documented yet.

    "style": {
      "description": "The file extension or preprocessor to use for style files.",
      "type": "string",
      "default": "css",
      "enum": [
        "css",
        "scss",
        "sass",
        "less",
        "styl"
      ],
      "x-prompt": {
        "message": "Which stylesheet format would you like to use?",
        "type": "list",
        "items": [
          { "value": "css",  "label": "CSS" },
          { "value": "scss", "label": "SCSS   [ https://sass-lang.com/documentation/syntax#scss                ]" },
          { "value": "sass", "label": "Sass   [ https://sass-lang.com/documentation/syntax#the-indented-syntax ]" },
          { "value": "less", "label": "Less   [ http://lesscss.org                                             ]" },
          { "value": "styl", "label": "Stylus [ http://stylus-lang.com                                         ]" }
        ]
      },
    },
</code-example>

#### x-prompt schema

The JSON schema that defines a schematic's options supports extensions to allow the declarative definition of prompts and their respective behavior.

> No additional logic or changes are required to the JavaScript for a schematic to support the prompts.

Reviewer 2: ```suggestion
No additional logic or changes are required to the code of a schematic to support the prompts.
```
Better to reduce possible confusion of JavaScript vs TypeScript here.

The following JSON schema is a complete description of the long-form syntax for the `x-prompt` field.

> <code-example language="json" header="x-prompt schema">

Reviewer 2: I was kind of hoping to find this somewhere in the source so that we could just reference it here instead of embedding it in the docs, but it looks like this is the only place that it exists. So I think we should keep it here.

Reviewer: It's here but outdated: https://github.com/angular/angular-cli/blob/master/packages/angular_devkit/core/src/json/schema/registry.ts#L614

There are three supported input types.

| Input type | Description |
| :----------- | :-------------------|
| confirmation | A yes or no question; ideal for Boolean options. |
| input | Textual input; ideal for string or number options. |
| list | A predefined set of allowed values. |

In the short form, the type is inferred from the property's type and constraints.

| Property Schema |	Prompt Type |
| :--------------- | :------------- |
| "type": "boolean" |	confirmation ("yes"=`true`, "no"=`false`) |
| "type": "string"  |	input |
| "type": "number"  |	input (only valid numbers accepted) |
| "type": "integer" |	input (only valid numbers accepted) |
| "enum": [...]   	| list 	(enum members become list selections) |

In the following example, the property takes an enumerated value, so the schematic automatically chooses the list type, and creates a menu from the possible values.

<code-example language="json" header="schema.json">

    "style": {
      "description": "The file extension or preprocessor to use for style files.",
      "type": "string",
      "default": "css",
      "enum": [
        "css",
        "scss",
        "sass",
        "less",
        "styl"
      ],

>       "x-prompt": ""Which stylesheet format would you like to use?"

Reviewer 2: ```suggestion
      "x-prompt": "Which stylesheet format would you like to use?"
```

Reviewer 2: Btw, thank you for adding this example. I think that it's very helpful!
